iati/core/rulesets.py

```diff
@@ -353,27 +461,51 @@ def is_valid_for(self, dataset):
 
         Raises:
             AttributeError: When an argument is given that does not have the required attributes.
-            XPathEvalError(lxml.etree.XPathEvalError): When no valid XPath is available.
+
+        Todo:
+            Determine if it's reasonable to assume the user should give a specific xpath format, or whether the context-path structure dictates automatic conversion to relative paths.
 
         """
-        found_paths = 0
+        def add_query_result(result):
+            """Add appropriate result to `found_in_dataset` whether attribute or element.
 
-        for path in self.paths:
-            result = dataset.xml_tree.xpath(path)
-            if result != list():
-                found_paths += 1
+            Args:
+                result (XPath element or attribute string): An XPath return value.
 
-        return not found_paths or found_paths == len(self.paths)
+            Todo:
+                Maybe refactor to return tag instead and extract to Rule base class.
+
+            """
+            try:
+                if result.is_attribute:
+                    found_in_dataset.add(result.getparent().tag)
+            except AttributeError:
+                found_in_dataset.add(result.tag)
+
+        context_elements = self._find_context_elements(dataset)
+        paths = set(self.paths)
+        found_in_dataset = set()
+
+        for context_element in context_elements:
+            if self._evaluate_condition(context_element):
+                return None
+            for path in paths:
+                results = context_element.xpath(path)
+                for result in results:
+                    if result != list():
```

Comment:
In what situation could `result` be a list, and what should happen in this situation?

Reply:
Added comment